File: check_braces.js

const fs = require('fs');

function checkBalance(filename) {
    try {
        const lines = fs.readFileSync(filename, 'utf-8').split('\n');
        const stack = [];
        const pairs = { ')': '(', '}': '{', ']': '[' };

        for (let lineNum = 0; lineNum < lines.length; lineNum++) {
            const line = lines[lineNum];
            for (let charPos = 0; charPos < line.length; charPos++) {
                const char = line[charPos];
                if ('({['.includes(char)) {
                    stack.push({ char, line: lineNum + 1, pos: charPos + 1 });
                } else if (')}]'.includes(char)) {
                    if (stack.length === 0) {
                        console.log(`Error: Unmatched '${char}' at line ${lineNum + 1}:${charPos + 1}`);
                        return;
                    }

                    const last = stack.pop();
                    if (last.char !== pairs[char]) {
                        console.log(`Error: Mismatched '${char}' at line ${lineNum + 1}:${charPos + 1}. Expected closing for '${last.char}' from line ${last.line}:${last.pos}`);
                        return;
                    }
                }
            }
        }

        if (stack.length > 0) {
            const last = stack[stack.length - 1];
            console.log(`Error: Unclosed '${last.char}' from line ${last.line}:${last.pos}`);
        } else {
            console.log("All braces balanced!");
        }
    } catch (err) {
        console.error("Error reading file:", err);
    }
}

checkBalance('src/app/configuracion/page.tsx');

File: check_braces.py

def check_balance(filename):
    with open(filename, 'r', encoding='utf-8') as f:
        lines = f.readlines()

    stack = []
    
    # Map of closing to opening chars
    pairs = {')': '(', '}': '{', ']': '['}
    
    for line_num, line in enumerate(lines, 1):
        for char_pos, char in enumerate(line, 1):
            if char in '({[':
                stack.append((char, line_num, char_pos))
            elif char in ')}]':
                if not stack:
                    print(f"Error: Unmatched '{char}' at line {line_num}:{char_pos}")
                    return
                
                last_char, last_line, last_pos = stack.pop()
                if last_char != pairs[char]:
                    print(f"Error: Mismatched '{char}' at line {line_num}:{char_pos}. Expected closing for '{last_char}' from line {last_line}:{last_pos}")
                    return

    if stack:
        char, line, pos = stack[-1]
        print(f"Error: Unclosed '{char}' from line {line}:{pos}")
    else:
        print("All braces balanced!")

check_balance('src/app/configuracion/page.tsx')
